src/fides/api/ops/service/connectors/saas/connector_registry_service.py

```diff
@@ -232,3 +286,74 @@ def update_saas_instance(
     connection_config.update_saas_config(db, SaaSConfig(**config_from_template))
 
     upsert_dataset_config_from_template(db, connection_config, template, template_vals)
+
+
+def register_custom_functions(script: str) -> None:
+    """
+    Registers custom functions by executing the given script in a restricted environment.
+
+    The script is compiled and executed with RestrictedPython, which is designed to reduce
+    the risk of executing untrusted code. It provides a set of safe builtins to prevent
+    malicious or unintended behavior.
+
+    Args:
+        script (str): The Python script containing the custom functions to be registered.
+
+    Raises:
+        SyntaxError: If the script contains a syntax error or uses restricted language features.
+        Exception: If an exception occurs during the execution of the script.
+    """
+
+    restricted_code = compile_restricted(
+        script, "<string>", "exec", policy=CustomRestrictingNodeTransformer
+    )
+    safe_builtins["__import__"] = custom_guarded_import
+    safe_builtins["_getitem_"] = getitem
+    safe_builtins["staticmethod"] = staticmethod
+
+    # pylint: disable=exec-used
+    exec(
+        restricted_code,
+        {
+            "__metaclass__": type,
+            "__name__": "restricted_module",
+            "__builtins__": safe_builtins,
+        },
+    )
+
+
+class CustomRestrictingNodeTransformer(RestrictingNodeTransformer):
+    """
+    Custom node transformer class that extends RestrictedPython's RestrictingNodeTransformer
+    to allow the use of type annotations (AnnAssign) in restricted code.
+    """
+
+    def visit_AnnAssign(self, node: AnnAssign) -> AST:
+        return self.node_contents_visit(node)
+
+
+def custom_guarded_import(
+    name: str,
+    _globals: Optional[dict] = None,
+    _locals: Optional[dict] = None,
+    fromlist: Optional[Tuple[str, ...]] = None,
+    level: int = 0,
+) -> Any:
+    """
+    A custom import function that prevents the import of certain potentially unsafe modules.
+    """
+    if name in [
+        "os",
+        "sys",
+        "subprocess",
+        "shutil",
+        "socket",
+        "importlib",
+        "tempfile",
+        "glob",
+    ]:
+        # raising SyntaxError to be consistent with exceptions thrown from other guarded functions
+        raise SyntaxError(f"Import of '{name}' module is not allowed.")
```

Comment:
a bit of a nit, but i do think it would be good to get some test coverage on this path. i'm not just saying it to be pedantic - it's an important security feature you've added